hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerMerger.java

```diff
@@ -0,0 +1,144 @@
+
+import static com.hazelcast.util.ExceptionUtil.rethrow;
+
+/**
+ * Merges data structures which have been collected via an {@link AbstractContainerCollector}.
+ *
+ * @param <C> container of the data structure
+ */
+public abstract class AbstractContainerMerger<C> implements Runnable {
+
+    private static final long TIMEOUT_FACTOR = 500;
+    private static final long MINIMAL_TIMEOUT_MILLIS = TimeUnit.SECONDS.toMillis(5);
+
+    protected final AbstractContainerCollector<C> collector;
+
+    private final Semaphore semaphore = new Semaphore(0);
+    private final ExecutionCallback<Object> mergeCallback = new ExecutionCallback<Object>() {
+        @Override
+        public void onResponse(Object response) {
+            semaphore.release(1);
+        }
+
+        @Override
+        public void onFailure(Throwable t) {
+            logger.warning("Error while running " + getLabel() + " merge operation: " + t.getMessage());
+            semaphore.release(1);
+        }
+    };
+
+    private final ILogger logger;
+    private final ClusterService clusterService;
+    private final OperationService operationService;
+    private final SplitBrainMergePolicyProvider splitBrainMergePolicyProvider;
+
+    private int operationCount;
+
+    protected AbstractContainerMerger(AbstractContainerCollector<C> collector, NodeEngine nodeEngine) {
+        this.collector = collector;
+        this.logger = nodeEngine.getLogger(AbstractContainerMerger.class);
+        this.clusterService = nodeEngine.getClusterService();
+        this.operationService = nodeEngine.getOperationService();
+        this.splitBrainMergePolicyProvider = nodeEngine.getSplitBrainMergePolicyProvider();
+    }
+
+    @Override
+    public final void run() {
+        // we cannot merge into a 3.9 cluster, since not all members may understand the new merge operation
+        // RU_COMPAT_3_9
+        if (clusterService.getClusterVersion().isLessThan(Versions.V3_10)) {
+            logger.info("Cluster needs to run version " + Versions.V3_10 + " to merge " + getLabel() + " instances");
+            return;
+        }
+        int valueCount = collector.getMergingValueCount();
+        if (valueCount == 0) {
+            return;
+        }
+
+        runInternal();
+
+        assert operationCount > 0 : "No merge operations have been invoked in AbstractContainerMerger";
+
+        try {
+            long timeoutMillis = Math.max(valueCount * TIMEOUT_FACTOR, MINIMAL_TIMEOUT_MILLIS);
+            if (!semaphore.tryAcquire(operationCount, timeoutMillis, TimeUnit.MILLISECONDS)) {
+                logger.warning("Split-brain healing for " + getLabel() + " didn't finish within the timeout...");
+            }
+        } catch (InterruptedException e) {
+            logger.finest("Interrupted while waiting for split-brain healing of " + getLabel() + "...");
+            Thread.currentThread().interrupt();
+        } finally {
+            collector.destroy();
```

Comment:
Is the operation serialized right away after the operation is submitted to the executor? I remember some kind of test failures in one of the many implementation, when I cleared the snapshot data immediately. But it could very well be the `OperationFactory` and `IMap`, not one of these data structures.

Since the node is a lite member until the merge operations are completed, it should not hurt to occupy the memory a tiny bit longer, so the `collector.destroy()` in the `finally` block just feels a bit safer for me.

Reply:
> Is the operation serialized right away after the operation is submitted to the executor?

Operations are [serialized just before being sent out over the network](https://github.com/hazelcast/hazelcast/blob/00ee8406b7945b601fc6c2e3783b7c53a5dd582d/hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/OutboundOperationHandler.java#L51) and it will be serialized again on each retry if it fails with retryable exception, so it's not safe to destroy the collector before operations are done.